import re
import subprocess as sp
from pathlib import Path

from snakemake.utils import available_cpu_count

IN_GENOMES = Path(config["genomes"])
IN_QUERIES_DIR = Path(config["queries"])
IN_BLAST_FIELDS = Path("config/blast_fields.txt")

assert IN_GENOMES.exists(), "Input genome assembly list file not found."
assert IN_QUERIES_DIR.exists(), "Input query directory not found."
assert IN_BLAST_FIELDS.exists(), "Input blast fields file not found."

RESULTS = Path(config["results"])
RESULTS_GENOMES = Path(f"{RESULTS}/genomes")


THREADS = (
    available_cpu_count() if config["threads"] == "all" else int(config["threads"])
)

GENOME_REGEX = r"GC[AF]_\d+\.\d"


wildcard_constraints:
    genome=GENOME_REGEX,


def fasta_entries(path):
    completed = sp.run(
        f"grep -c '^>' {path}", shell=True, capture_output=True, check=True
    )
    return int(completed.stdout)


def parse(path, mode=None):
    with open(path, "r") as file:
        out = list()
        for line in file:
            line = re.sub(r"#.*$", "", line)  # rm comments
            line = line.strip()  # rm leading and trailing whitespace
            if line == "":
                continue
            if mode != "genome":
                out.append(line)
            elif re.match(GENOME_REGEX, line).group(0) == line:
                out.append(line)  # append only valid assembly ids
    return out


GENOMES = parse(IN_GENOMES, mode="genome")
BLAST_FIELDS = " ".join(parse(IN_BLAST_FIELDS))


rule help:
    shell:
        "bat workflow/Snakefile"


rule download_genome:
    output:
        faa=f"{RESULTS_GENOMES}/{{genome}}/{{genome}}.faa",
        gff=f"{RESULTS_GENOMES}/{{genome}}/{{genome}}.gff",
    shell:
        """
        mkdir -p {RESULTS_GENOMES}/{wildcards.genome}
        workflow/scripts/download_genome.py --include protein gff3 --out-dir {RESULTS_GENOMES}/{wildcards.genome} -- {wildcards.genome}
        """


rule blastp:
    input:
        faa=rules.download_genome.output.faa,
        query=f"{IN_QUERIES_DIR}/{{query}}.faa",
    output:
        tsv=f"{RESULTS_GENOMES}/{{genome}}/{{genome}}_{{query}}.tsv",
    params:
        db=f"{RESULTS_GENOMES}/{{genome}}/{{genome}}",
        format=f"{BLAST_FIELDS}",
    shell:
        """
        diamond makedb --in {input.faa} -d {params.db}
        diamond blastp --outfmt {params.format}\
            -o {output.tsv}\
            -d {params.db}\
            -q {input.query}
        """
